core/src/main/java/feast/core/service/JobCoordinatorService.java

```diff
@@ -16,196 +16,136 @@
+        }
+        if (!featureSetSpecs.isEmpty()) {
           featureSetSpecs.stream()
-              .map(fs -> fs.getName() + ":" + fs.getVersion())
-              .collect(Collectors.toSet());
-      if (existingFeatureSetsPopulatedByJob.size() == newFeatureSetsPopulatedByJob.size()
-          && existingFeatureSetsPopulatedByJob.containsAll(newFeatureSetsPopulatedByJob)) {
-        return job.get();
-      } else {
-        return updateJob(job.get(), featureSetSpecs, store);
+              .collect(Collectors.groupingBy(FeatureSetSpec::getSource))
+              .entrySet()
+              .stream()
+              .forEach(
+                  kv -> {
+                    Optional<JobInfo> originalJob =
+                        getJob(Source.fromProto(kv.getKey()), Store.fromProto(store));
+                    jobUpdateTasks.add(
+                        new JobUpdateTask(
+                            kv.getValue(), kv.getKey(), store, originalJob, jobManager));
+                  });
+        }
+      } catch (InvalidProtocolBufferException e) {
+        log.warn("Unable to retrieve feature sets for store {}: {}", store, e.getMessage());
       }
-    } else {
-      return startJob(
-          createJobId(source.getId(), store.getName()), featureSetSpecs, sourceSpec, store);
     }
-  }
-
-  /** Get the non-terminal job associated with the given featureSet name and store name, if any. */
-  private Optional<JobInfo> getJob(String sourceId, String storeName) {
-    List<JobInfo> jobs = jobInfoRepository.findBySourceIdAndStoreName(sourceId, storeName);
-    if (jobs.isEmpty()) {
-      return Optional.empty();
+    if (jobUpdateTasks.size() == 0) {
+      log.info("No jobs found.");
+      return;
     }
-    return jobs.stream()
-        .filter(job -> !(JobStatus.getTerminalState().contains(job.getStatus())))
-        .findFirst();
-  }
-
-  /** Start or update the job to ingest data to the sink. */
-  private JobInfo startJob(
-      String jobId,
-      List<FeatureSetSpec> featureSetSpecs,
-      SourceProto.Source source,
-      StoreProto.Store sinkSpec) {
-    try {
-      AuditLogger.log(
-          Resource.JOB,
-          jobId,
-          Action.SUBMIT,
-          "Building graph and submitting to %s",
-          jobManager.getRunnerType().getName());
 
-      String extId = jobManager.startJob(jobId, featureSetSpecs, sinkSpec);
-      if (extId.isEmpty()) {
-        throw new RuntimeException(
-            String.format("Could not submit job: \n%s", "unable to retrieve job external id"));
+    log.info("Creating/Updating {} jobs...", jobUpdateTasks.size());
+    ExecutorService executorService = Executors.newFixedThreadPool(jobUpdateTasks.size());
+    ExecutorCompletionService<JobInfo> ecs = new ExecutorCompletionService<>(executorService);
+    jobUpdateTasks.forEach(ecs::submit);
+
+    int completedTasks = 0;
+    while (completedTasks < jobUpdateTasks.size()) {
+      try {
+        JobInfo jobInfo = ecs.take().get();
```

Comment:
No, timeouts are implemented inside the task execution itself, since timeouts at the ecs level are incremental (e.g. timeout of 4 min means up to 4 min for the first task that arrives, up to 8 min for the second task that arrives, etc.)